scipy/optimize/_minimize.py

```diff
@@ -531,29 +535,25 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
              RuntimeWarning)
 
     # check gradient vector
-    if meth == 'trust-constr':
-        if type(jac) is bool:
-            if jac:
-                fun = MemoizeJac(fun)
-                jac = fun.derivative
-            else:
-                jac = '2-point'
-        elif jac is None:
-            jac = '2-point'
-        elif not callable(jac) and jac not in ('2-point', '3-point', 'cs'):
-            raise ValueError("Unsupported jac definition.")
+    if callable(jac):
+        pass
+    elif jac is True:
+        # fun returns func and grad
+        fun = MemoizeJac(fun)
+        jac = fun.derivative
+    elif (jac in ('2-point', '3-point', 'cs') and
+          meth in ['trust-constr', 'bfgs', 'cg', 'l-bfgs-b', 'tnc']):
+        # finite differences
+        pass
+    elif meth in ['trust-constr']:
```

Comment:
It has been a long time since I last looked into that. But I understood that '2-point' and None is the same in the end. 

Could we colapse the last 2 elif and the last else into just:
```
else:
    jac = None
```
??

Reply:
Because it seems to me that the result would be the same as:
```
elif meth in ['trust-constr']:
       # default jac calculation for this method
        jac = '2-point'
elif jac is None or bool(jac) is False:
    # this will cause e.g. LBFGS to use forward difference, absolute step
    jac = None
else
    # default if jac option is not understood
     jac = None
```

Reply:
>  But I understood that '2-point' and None is the same in the end.

You're correct for the existing behaviour. Before this PR if the method was `'2-point'` and the method wasn't 'trust-constr' `jac` would be [converted](https://github.com/scipy/scipy/blob/master/scipy/optimize/_minimize.py#L550) to `None`. 
In all the `_minimize_*` methods that would correspond to using forward differences with an *absolute* step using `approx_fprime`. 
If the 'trust_constr' method was used `jac in {None, '2-point'}` would be done via forward differences with a *relative* step in `approx_derivative`.

In this PR we want to ensure back compatibility. So if:
- `jac=None` the gradient will still be calculated using forward differences with an *absolute* step in `approx_derivative`.
- `jac=None` and `meth='trust-constr`, still needs forward differences with a *relative* step, so make `jac='2-point'`. If `None` is sent to 'trust-constr' then the method won't understand it (see [here](https://github.com/scipy/scipy/blob/master/scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py#L331), then [here](https://github.com/scipy/scipy/blob/master/scipy/optimize/_differentiable_functions.py#L31))
- The `else` is equivalent to arriving at this [line](https://github.com/scipy/scipy/blob/master/scipy/optimize/_minimize.py#L556). I think I tried removing the else and there were test failures.

The new behaviour is that minimizer methods can now use `'2-point', '3-point', 'cs'`, via `approx_derivative` and a *relative step*. In other words:
- `jac=None`-->absolute step with forward differences
- `jac in {'2-point', '3-point', 'cs'}` --> relative step. 
Relative step is probably better in nearly all cases. If the latter is employed then a newly added `finite_diff_rel_step` keyword is used to figure out what the relative step is. This keyword naming followed 'trust-constr'.

Currently there is no way of the minimizer methods asking for `2-point`, `3-point`, or `cs` with an absolute step. This would require an extra keyword of the form `step_type={'relative', 'absolute'}`. I figured this change could be made in the future if absolutely required, but at the expense of adding another keyword.
(note that `approx_derivative` could handle either calculation).

Reply:
Ahh ok. Good that you considered that.

You can probably merge the else with last elif though, since both set `jac = None`